codepropertygraph/src/main/scala/io/shiftleft/utils/TimeMetric.scala

```diff
@@ -0,0 +1,349 @@
+    recorder match {
+      case Some(rec) =>
+        rec.pushEvent(
+          s"${getNewTime()}, $message, Total Time taken, ${getTheTotalTime()}, ${String
+              .format("%.2f", osBean.getProcessCpuLoad * 100)}, ${String.format("%.2f", usedMemory)}, ${String.format("%.2f", maxHeap)}"
+        )
+        Thread.sleep(100)
+        rec.interrupt()
+      case _ =>
+    }
+  }
+
+  private def pushToRecorder(
+    event: String,
+    stage: String,
+    status: String,
+    usedMemory: Double,
+    maxHeap: Double,
+    cpuUsed: Double
+  ): Unit = {
+    recorder match {
+      case Some(rec) =>
+        rec.pushEvent(s"$event, $status, ${String
+            .format("%.2f", cpuUsed)}, ${String.format("%.2f", usedMemory)}, ${String.format("%.2f", maxHeap)}")
+        status match {
+          case STARTED => stagePerformance += (stage -> NumberProcessor())
+          case _       =>
+        }
+      case _ =>
+    }
+  }
+
+  private def initiateNewSubStage(subStageName: String, additionalMetaDataToLog: String = STAGE_NOT_SET): Unit = {
+    val (usedMemory, maxHeap, cpuUsed, _) = getCurrentCpuMemoryStats()
+    pushToRecorder(
+      s"${getNewTime()}, $lastStageName, $subStageName",
+      s"-- $subStageName",
+      STARTED,
+      usedMemory,
+      maxHeap,
+      cpuUsed
+    )
+    val formattedString =
+      s"%-${maxStartingBuffer}s%s"
+        .format(
+          s"- --$subStageName is invoked",
+          s"- stats -> ${currentCpuMemoryStatsInString(cpuUsed, usedMemory, maxHeap)}"
+        )
+    logger(s"${getNewTime()} $formattedString", true)
+    lastSubStageName = subStageName
+    additionalMetaData = additionalMetaDataToLog
+  }
+
+  private def endLastSubStage(): Unit = {
+    val (usedMemory, maxHeap, cpuUsed, maxStats) = getCurrentCpuMemoryStats(Some(s"-- $lastSubStageName"))
+    pushToRecorder(
+      s"${getNewTime()}, $lastStageName, $lastSubStageName",
+      lastSubStageName,
+      DONE,
+      usedMemory,
+      maxHeap,
+      cpuUsed
+    )
+    val formattedString = s"%-${maxEndingBuffer}s%s".format(
+      s"- --$lastSubStageName is done ...",
+      s"%-${maxBuffers.MAX_DONE_TIME_BUFFER}s%s".format(
+        s"- ${setNewTimeToLastAndGetTimeDiff()}",
+        s"- stats -> ${currentCpuMemoryStatsInString(cpuUsed, usedMemory, maxHeap, maxStats)}"
+      )
+    )
+    logger(s"${getNewTime()} $formattedString", true)
+    lastSubStageName = STAGE_NOT_SET
+    additionalMetaData = STAGE_NOT_SET
+  }
+
+  private def getCurrentCpuMemoryStats(
+    stage: Option[String] = None
+  ): (Double, Double, Double, Option[(Double, Double)]) = {
+    stage match {
+      case Some(st) if stagePerformance.contains(st) =>
+        val numberProcessor     = stagePerformance(st)
+        val (avgCpu, avgMemory) = numberProcessor.getAverage
+        val (maxCpu, maxMemory) = numberProcessor.getMax
+        (avgMemory, Runtime.getRuntime().maxMemory() * 0.000001, avgCpu, Some(maxCpu, maxMemory))
+      case _ =>
+        (
```

Comment:
Could you create a case class instead of return all of this in a tuple, please? It would help make it more descriptive as to what each variable is

Reply:
done